### ISSUE TITLE
Prefer `[` over "[" for prefix form of non-syntactic functions

### ISSUE THREAD

**Issue by MichaelChirico** [r-lib/lintr#1444]
Just saw this clean-up in r-devel was done:

https://github.com/r-devel/r-svn/commit/ff5193fd1f4d17327a56858745ae14df0d1f4930

Using backticks avoids a `match.fun()` call, so should be a mite faster.

Personally, I've also always found the use of `"` a bit confusing -- I think `` `[` `` makes the intent clearer for less experienced readers.

**Comment by IndrajeetPatil** [comment]
Do you mean we should introduce a new linter that makes this suggestion (prefer `` `[` `` over `"["`), or that we should do a similar clean-up in `{lintr}` source code?

If it's the latter, I couldn't find any instance of usage of `"["` in a function call.

**Comment by MichaelChirico** [comment]
the former (and the latter since we like to apply our own linters)

**Comment by Bisaloo** [comment]
Should this be part of a more generic linter to prevent passing functions as strings?

See [this recent comment from Martin Mächler](https://bugs.r-project.org/show_bug.cgi?id=18429#c1) saying passing functions as symbols is preferred. There is also a [relevant twitter thread](https://twitter.com/henrikbengtsson/status/1579871758081220611) where opinions are more diverse.

**Comment by Bisaloo** [comment]
> Should this be part of a more generic linter to prevent passing functions as strings?

Is this `KeywordQuoteLinter` from https://github.com/r-lib/lintr/issues/884?

**Comment by MichaelChirico** [comment]
This issue is _partially_ covered by the new `keyword_quote_linter()`. What remains is if we'd like to look for this in cases not covered by `keyword_quote_linter()`, e.g.

```r
sapply(x, "[", 1)
```

`keyword_quote_linter()` right now has no way of knowing `"["` is actually a function name there.

So we could extend `keyword_quote_linter()` with a few rules:

 - `STR_CONST` in the `n`th argument of common mapping functions like `sapply()`, `vapply()` `lapply()`, `tapply()`, `rapply()`, `mapply()`, `Map()`, ... ?
 - Any code like `FUN = "["`, i.e. the argument `FUN` is always assumed to be for function names

Does that make sense?

**Comment by MichaelChirico** [comment]
Copying over a comment from #2280 to make sure `do.call()` is also linted by this issue.

> I recently read some discussions about the differences between `do.call("rbind", ...)` and `do.call(rbind, ...)`. R-core is [changing some internal code](https://github.com/r-devel/r-svn/commit/786d5a44a338c71d09b8e1ee32ef5fbabb8ef1ec) to use the later.
> 
> Similarly, there is [another change](https://github.com/r-devel/r-svn/commit/3cabcd4fd011e95e3f9de405d91c21f2653d4efc) replacing some calls to "[" to "`[`" to avoid an internal call to `match.fun()`.
> 
> I couldn't find a linter for this; is this something that could be useful? This linter might be under[ "Efficiency linters"](https://lintr.r-lib.org/reference/efficiency_linters.html)

**Comment by AshesITR** [comment]
Just an idea: Do a reverse search for calls to `match.fun()` from base and recommended packages to make sure we find all relevant functions.
